osgp/platform/osgp-adapter-kafka-distributionautomation/src/test/java/org/opensmartgridplatform/adapter/kafka/da/infra/kafka/out/MeterReadingProducerTest.java

```diff
@@ -0,0 +1,104 @@
+import org.opensmartgridplatform.adapter.kafka.da.avro.UsagePoint;
+import org.opensmartgridplatform.adapter.kafka.da.avro.ValuesInterval;
+import org.opensmartgridplatform.domain.da.measurements.MeasurementReport;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.kafka.core.ConsumerFactory;
+import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
+import org.springframework.kafka.core.KafkaTemplate;
+import org.springframework.kafka.test.EmbeddedKafkaBroker;
+import org.springframework.kafka.test.context.EmbeddedKafka;
+import org.springframework.kafka.test.utils.KafkaTestUtils;
+import org.springframework.test.context.TestPropertySource;
+import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
+
+@SpringJUnitConfig(KafkaProducerConfig.class)
+@TestPropertySource("classpath:osgp-adapter-kafka-distributionautomation-test.properties")
+@ExtendWith(MockitoExtension.class)
+@EmbeddedKafka(partitions = 1,
+        topics = { "${distributionautomation.kafka.producer.topic}" },
+        brokerProperties = { "listeners=PLAINTEXT://localhost:9092", "log.dirs=../kafka-logs/",
+                "auto.create.topics.enable=true" })
+class MeterReadingProducerTest {
+
+    @Value("${distributionautomation.kafka.producer.topic}")
+    private String topic;
+
+    @Autowired
+    private EmbeddedKafkaBroker embeddedKafka;
+
+    @Mock
+    private DistributionAutomationMapper mapper;
+
+    @Autowired
+    private KafkaTemplate<String, MeterReading> template;
+
+    private MeterReadingProducer producer;
+    private MeterReading message;
+
+    @BeforeEach
+    @SuppressWarnings("unchecked")
+    public void setup() {
+        this.message = this.createMessage();
+        when(this.mapper.map(any(MeasurementReport.class), any(Class.class))).thenReturn(this.message);
+        this.producer = new MeterReadingProducer(this.template, this.mapper);
+    }
+
+    @Test
+    void sendTest() {
+
+        // send a message to the kafka bus
+        this.producer.send(new MeasurementReport.Builder().build());
+
+        // consume the message with embeddedKafka
+        final Map<String, Object> consumerProps = KafkaTestUtils.consumerProps("testGroup", "true", this.embeddedKafka);
+        consumerProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        final ConsumerFactory<String, MeterReading> consumerFactory = new DefaultKafkaConsumerFactory<>(consumerProps,
+                new StringDeserializer(), new MeterReadingDeserializer());
+        final Consumer<String, MeterReading> consumer = consumerFactory.createConsumer();
+        this.embeddedKafka.consumeFromAnEmbeddedTopic(consumer, this.topic);
+        final ConsumerRecord<String, MeterReading> received = KafkaTestUtils.getSingleRecord(consumer, this.topic);
+
+        // check the consumed message
+        assertThat(received).has(value(this.message));
```

Comment:
When we receive a message, it is always mapped to the result of createMessage(), regardless of the message sent by the producer. Was this intended by this test?
With some more work, we could construct a specific MeasurementReport, which maps to the result of createMessage().

Reply:
I think that mapping should be tested in another class.